src/compiler/checker.ts

```diff
@@ -3904,14 +3907,18 @@ namespace ts {
                 }
 
                 function typeReferenceToTypeNode(type: TypeReference) {
-                    const typeArguments: ReadonlyArray<Type> = type.typeArguments || emptyArray;
+                    const typeArguments: ReadonlyArray<Type> = getTypeArguments(type);
                     if (type.target === globalArrayType || type.target === globalReadonlyArrayType) {
                         if (context.flags & NodeBuilderFlags.WriteArrayAsGenericType) {
-                            const typeArgumentNode = typeToTypeNodeHelper(typeArguments[0], context);
-                            return createTypeReferenceNode(type.target === globalArrayType ? "Array" : "ReadonlyArray", [typeArgumentNode]);
+                            depth++;
+                            const typeArgumentNode = typeToTypeNodeHelper(depth >= 5 ? anyType : typeArguments[0], context);

```

Comment:
We don't use straight `any` in situations like this in the node builder. Use `createElidedInformationPlaceholder(context)` (which appropriately swaps between `any` and the slightly more helpful and correct `...` based on context) instead~

Also this depth is a bad heuristic and will break declaration files - since it's counting global nested type references, a `A<B<C<D<E<F<whatever>>>>>>` won't serialize without deleting information, which is hugely problematic. The depth needs to be tracked per-symbol, at least (and ideally only per-type, if viable). There's already a map in `context` for this - `context.visitedTypes` for tracking exact types seen (which should be set on enter and unset on exit), and `context.symbolDepth` for tracking how many times we've visited a single symbol in the current stack (which should be incremented on enter and decremented on exit). The logic to handle them that's currently in `createAnonymousTypeNode` should probably just be extracted and reused here.